sw/host/ot_certs/src/template/mod.rs

```diff
@@ -0,0 +1,537 @@
+
+mod hjson;
+
+/// Full template file, including variable declarations and certificate spec.
+#[serde_as]
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq)]
+pub struct Template {
+    /// Name of the certificate.
+    pub name: String,
+    /// Variable declarations.
+    pub variables: HashMap<String, VariableType>,
+    /// Certificate specification.
+    pub certificate: Certificate,
+}
+
+/// Certificate specification.
+#[serde_as]
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize)]
+pub struct Certificate {
+    /// X509 certificate's serial number
+    #[serde_as(as = "Value<hjson::BigUint>")]
+    pub serial_number: Value<BigUint>,
+    /// X509 certificate's issuer.
+    pub issuer: HashMap<AttributeType, Value<String>>,
+    /// X509 certificate's subject.
+    pub subject: HashMap<AttributeType, Value<String>>,
+    /// X509 certificate's public key.
+    pub subject_public_key_info: SubjectPublicKeyInfo,
+    /// X509 certificate's authority key identifier.
+    #[serde_as(as = "Value<hjson::HexString>")]
+    pub authority_key_identifier: Value<Vec<u8>>,
+    /// X509 certificate's public key identifier.
+    #[serde_as(as = "Value<hjson::HexString>")]
+    pub subject_key_identifier: Value<Vec<u8>>,
+    /// DICE TCB model.
+    pub model: Option<Value<String>>,
+    /// DICE TCB vendor.
+    pub vendor: Option<Value<String>>,
+    /// DICE TCB version.
+    pub version: Option<Value<String>>,
+    /// DICE TCB security version number.
+    #[serde_as(as = "Option<Value<hjson::BigUint>>")]
+    pub svn: Option<Value<BigUint>>,
+    /// DICE TCB layer.
+    #[serde_as(as = "Option<Value<hjson::BigUint>>")]
+    pub layer: Option<Value<BigUint>>,
+    /// DICE TCB firmware IDs.
+    pub fw_ids: Option<Vec<FirmwareId>>,
+    /// DICE TCB flags.
+    pub flags: Option<Flags>,
+    /// X509 certificate's signature.
+    pub signature: Signature,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Hash, strum::Display)]
+#[serde(rename_all = "snake_case")]
+pub enum AttributeType {
+    #[serde(alias = "c")]
+    Country,
+    #[serde(alias = "o")]
+    Organization,
+    #[serde(alias = "ou")]
+    OrganizationalUnit,
+    #[serde(alias = "st")]
+    State,
+    #[serde(alias = "cn")]
+    CommonName,
+    #[serde(alias = "sn")]
+    SerialNumber,
+}
+
+/// Value which may either be a variable name or literal.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Value<T> {
+    /// This value will be populated on the device when variables are set.
+    Variable {
+        /// Name of the variable.
+        var: String,
+        /// Optional conversion to apply to the variable.
+        convert: Option<Conversion>,
+    },
+    /// Constant literal that will be set when the certificate is generated.
+    Literal(T),
+}
+
+impl<T> Value<T> {
+    /// Create a variable with the given name. No conversion applied.
+    pub fn variable(name: &str) -> Self {
+        Value::Variable {
+            var: String::from(name),
```

Comment:
`name.to_owned()` if you want to be explicit, or `name.into()` to use the implicit form.

Reply:
Do you know if there's a most commonly used one? I think I use `String::from` and `.to_string()` most often, but are these preferred?

Reply:
Historical `to_owned()` is preferred because it's more explicit than `From/Into` and more performant than `to_string()` (note that this calls the `Display` impl of `&str` so conceptually have to go through the formatting machinery). But nowadays specialisation is added in std so that `to_string()` is identical to `to_owned()`.

Many people (myself included) still prefer semantics of getting an owned copy (and `to_owned()` works on both string and slice), but some people prefer `to_vec()` and `to_string()`. I'd say either `to_owned()` and `to_string()` is fine.